### PR TITLE
Make it optional to subtract mean pixel or mean image

### LINKED ISSUE #169: Subtract mean image or pixel in both training and inference
Thanks to Yuguang Lee for pointing this out to me on [digits-users](https://groups.google.com/d/msg/digits-users/WRAuVqnyi3g/Fw8NODUBDLQJ).

## Problem

Currently, DIGITS subtracts the mean **image** during training (see [here](https://github.com/NVIDIA/DIGITS/blob/v2.0.0-preview/digits/model/tasks/caffe_train.py#L232-L235)) and the mean **pixel** during inference (see [here](https://github.com/NVIDIA/DIGITS/blob/v2.0.0-preview/digits/model/tasks/caffe_train.py#L963-L972)). That's both illogical and incorrect [but still works pretty well, which is why I hadn't noticed it].

## Explanation

I moved to subtracting the mean pixel during inference because it simplifies the inference path. I'll give an example to explain:

Let's say you have 256x256 images in your dataset and your crop size is 227. During training, Caffe first subtracts the 256x256 mean image from the input image, and then takes a random 227x227 crop of the image before passing it to your network. But at inference time, the network only knows that images are supposed to be 227x227. So DIGITS resizes your test image to 227x227 and then subtracts the mean pixel across the whole image before passing it to the network.

## Solution

DIGITS should allow you to do either mean pixel subtraction or mean image subtraction, but it shouldn't do this strange hybrid.

####  Mean pixel

1. For training - provide `transform_param.mean_value` (see [here](https://github.com/NVIDIA/caffe/blob/v0.12.1/src/caffe/proto/caffe.proto#L356-L359)).
2. For inference -  resize test image to crop size and then subtract the mean pixel. DIGITS already does this.

#### Mean image

1. For training - provide `transform_param.mean_file` (see [here](https://github.com/NVIDIA/caffe/blob/v0.12.1/src/caffe/proto/caffe.proto#L354-L355)). DIGITS already does this.
2. For inference - resize test image to the mean file size, subtract mean, resize **again** to the crop size.

### PR DESCRIPTION
**Pull request by jmancewicz**
Closes #169

### CODE REVIEW

File: digits/model/tasks/caffe_train.py [hunk 1]
```diff
@@ -1434,9 +1465,12 @@ def get_transformer(self):
...
             t.set_channel_swap('data', channel_swap)
 
-        if mean_pixel is not None:
-            # set mean
+        # set mean
+        if self.use_mean == 'pixel' and mean_pixel is not None:
             t.set_mean('data', mean_pixel)
+        elif self.use_mean == 'image':
+            mean_image =  self.get_mean_image(mean_file)
+            t.set_mean('data', mean_image)
```
Comment (lukeyeager) [inline digits/model/tasks/caffe_train.py]: I think you need to crop the image here.
```
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/lib/python2.7/dist-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/lyeager/digits/digits/model/images/classification/views.py", line 276, in image_classification_model_classify_one
    predictions, visualizations = job.train_task().infer_one(image, snapshot_epoch=epoch, layers=layers)
  File "/home/lyeager/digits/digits/model/tasks/caffe_train.py", line 972, in infer_one
    layers=layers,
  File "/home/lyeager/digits/digits/model/tasks/caffe_train.py", line 1002, in classify_one
    preprocessed = self.get_transformer().preprocess(
  File "/home/lyeager/digits/digits/model/tasks/caffe_train.py", line 1473, in get_transformer
    t.set_mean('data', mean_image)
  File "/home/lyeager/caffe/nv-0.13.2/python/caffe/io.py", line 255, in set_mean
    raise ValueError('Mean shape incompatible with input shape.')
ValueError: Mean shape incompatible with input shape.
```

**Comment by jmancewicz** [comment]
get_mean_image does an imresize.  I'll look into what's going wrong.

**Comment by jmancewicz** [comment]
@gheinrich, please look at the changes to torch_train.py and see if they are correct.

**Comment by gheinrich** [comment]
@jmancewicz the parameter mappings look OK to me. I have trained LeNet on MNIST using all three methods (subtractMeanImage, subtractMeanPixel, none) and they all converged in approximately the same time. Maybe MNIST is too trivial to actually observe the effect of mean subtraction... anyway, the Torch part looks OK to me thanks!

**Comment by lukeyeager** [comment]
You're getting a bunch of test failures.

https://travis-ci.org/NVIDIA/DIGITS/builds/83967654
https://s3.amazonaws.com/archive.travis-ci.org/jobs/83967655/log.txt

There are a bunch of weird PIL messages too - any idea what that's about?

    PIL.PngImagePlugin: DEBUG: STREAM IHDR 16 13

**Comment by gheinrich** [comment]
@jmancewicz the support for generic inference with Torch is now on the master branch. There might be some rebasing to do for this PR. Let me know when you have rebased as I now have a network for which mean subtraction is critical so I will be able to test how the network behaves depending on the mean subtraction method.

**Comment by lukeyeager** [comment]
@jmancewicz  can you make sure that #365 is fixed in this PR too?

**Comment by jmancewicz** [comment]
To handle #365, I'm deleting setting mean_pixel rather than appending, and deleting mean_file when using mean_pixel.  Using mean mean_file will similarly del mean_pixel and set mean_file.

```
        if self.use_mean == 'pixel':
            assert dataset_backend != 'hdf5', 'HDF5Data layer does not support mean subtraction'
            mean_pixel = self.get_mean_pixel(self.dataset.path(self.dataset.train_db_task().mean_file))
            ## remove any values that may already be in the network
            if hasattr(train_data_layer.transform_param, 'mean_file'):
                del(train_data_layer.transform_param.mean_file)
            train_data_layer.transform_param.mean_value = mean_pixel[:]
            if val_data_layer is not None and has_val_set:
                if hasattr(val_data_layer.transform_param, 'mean_file'):
                    del(val_data_layer.transform_param.mean_file)
                val_data_layer.transform_param.mean_value = mean_pixel[:]
        elif self.use_mean == 'image':
            ## remove any values that may already be in the network
            if hasattr(train_data_layer.transform_param, 'mean_pixel'):
                del(train_data_layer.transform_param.mean_pixel)
            train_data_layer.transform_param.mean_file = self.dataset.path(self.dataset.train_db_task().mean_file)
            if val_data_layer is not None and has_val_set:
                if hasattr(val_data_layer.transform_param, 'mean_pixel'):
                    del(val_data_layer.transform_param.mean_pixel)
                val_data_layer.transform_param.mean_file = self.dataset.path(self.dataset.train_db_task().mean_file)
```

File: digits/model/tasks/torch_train.py [hunk 2]
```diff
@@ -136,6 +135,8 @@ def task_arguments(self, resources, env):
                 '--save=%s' % self.job_dir,
                 '--snapshotPrefix=%s' % self.snapshot_prefix,
                 '--snapshotInterval=%s' % self.snapshot_interval,
+                '--mean=%s' % self.dataset.path(constants.MEAN_FILE_IMAGE),
+                '--labels=%s' % self.dataset.path(self.dataset.labels_file),
```
Comment (lukeyeager) [inline digits/model/tasks/torch_train.py]: Datasets don't necessarily have a `labels_file`. That's why you're getting these TravisCI errors for generic Torch tests:
```
======================================================================
FAIL: digits.model.images.generic.test_views.TestTorchCreation.test_create_wait_delete
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/NVIDIA/DIGITS/digits/model/images/generic/test_views.py", line 245, in test_create_wait_delete
    assert self.model_wait_completion(job_id) == 'Done', 'create failed'
AssertionError: create failed
-------------------- >> begin captured logging << --------------------
digits: ERROR: AttributeError: 'GenericImageDatasetJob' object has no attribute 'labels_file'
--------------------- >> end captured logging << ---------------------
```

Comment (lukeyeager) [inline digits/model/tasks/torch_train.py]: Why are you adding labels at all? [It's already being done below](https://github.com/NVIDIA/DIGITS/blob/13d28b4079eead5e9a3b2afadf1806d3f0af0834/digits/model/tasks/torch_train.py#L150).

File: digits/model/tasks/torch_train.py [hunk 3]
```diff
@@ -136,6 +135,8 @@ def task_arguments(self, resources, env):
                 '--save=%s' % self.job_dir,
                 '--snapshotPrefix=%s' % self.snapshot_prefix,
                 '--snapshotInterval=%s' % self.snapshot_interval,
+                '--mean=%s' % self.dataset.path(constants.MEAN_FILE_IMAGE),
```
Comment (lukeyeager) [inline digits/model/tasks/torch_train.py]: Why are you adding the mean here? [It's already being done below](https://github.com/NVIDIA/DIGITS/blob/13d28b4079eead5e9a3b2afadf1806d3f0af0834/digits/model/tasks/torch_train.py#L149).

File: digits/model/tasks/torch_train.py (hunk 2 above)
Comment (gheinrich) [inline digits/model/tasks/torch_train.py]: these two lines were moved to lines 150-151
line 192 addresses the generic dataset case (where we need to create an image file out of the .protobuf from the dataset)

File: digits/model/tasks/torch_train.py [hunk 4]
```diff
@@ -829,14 +843,24 @@ def classify_many(self, images, snapshot_epoch=None):
...
+                args.append('--useMeanPixel=yes')
+                args.append('--subtractMean=no')
+            elif self.use_mean == 'image':
+                args.append('--useMeanPixel=no')
+                args.append('--subtractMean=yes')
+            else:
+                args.append('--useMeanPixel=no')
+                args.append('--subtractMean=no')
+
+            #print args
```
Comment (gheinrich) [inline digits/model/tasks/torch_train.py]: while you're at it, you might want to remove this comment :-)

File: digits/model/tasks/torch_train.py (hunk 2 above)
Comment (gheinrich) [inline digits/model/tasks/torch_train.py]: line 174:
```
if self.use_mean:
```
Maybe this should be:
```
if self.use_mean != 'none':
```

**Comment by lukeyeager** [comment]
Looks like I added a bug way back in #41. At [line 1427](https://github.com/NVIDIA/DIGITS/blob/v2.2.1/digits/model/tasks/caffe_train.py#L1427), this:
```python
t = caffe.io.Transformer(
    inputs = {'data':  data_shape}
    )
```
needs to change to this.
```python
t = caffe.io.Transformer(
    inputs = {'data':  tuple(data_shape)}
    )
```
It was originally me that added the overly strict check to Caffe with https://github.com/BVLC/caffe/pull/2031. Now I wish I would have done it in such a way that it worked with either `lists` or `tuples`.

We didn't notice the issue in DIGITS because we haven't done mean file subtraction for so long.

File: digits/model/tasks/caffe_train.py [hunk 5]
```diff
@@ -118,6 +119,70 @@ def before_run(self):
...
+    def set_mean_value(self, layer, mean_pixel):
+        # remove any values that may already be in the network
+        layer.transform_param.ClearField('mean_file')
+        layer.transform_param.ClearField('mean_value')
+        layer.transform_param.mean_value.extend(list(mean_pixel))
+
+    def set_mean_file(self, layer, mean_file):
+        # remove any values that may already be in the network
+        layer.transform_param.ClearField('mean_value')
+        layer.transform_param.mean_file = mean_file
```
Comment (lukeyeager) [inline digits/model/tasks/caffe_train.py]: These two functions do the correct thing from a functional point of view (at least I think they do - haven't tested it), but I'd like to print some WARNINGs to the log if any of these `ClearField()` calls are actually overwriting any information.

**Comment by jmancewicz** [comment]
@lukeyeager I've added tests for mean image and pixel for training and inference.  Also dependency on the deploy.prototxt was avoidable, because I didn't need to imgresize the mean file at that time.

File: digits/model/images/classification/test_views.py [hunk 6]
```diff
@@ -317,6 +317,20 @@ def check_select_gpus(self, gpu_list):
         job_id = self.create_model(select_gpus_list=','.join(gpu_list), batch_size=len(gpu_list))
         assert self.model_wait_completion(job_id) == 'Done', 'create failed'
 
+    def test_mean_image(self):
+        options = {
+            'use_mean': 'image',
```
Comment (gheinrich) [inline digits/model/images/classification/test_views.py]: if my understanding is correct the default value for `use_mean` is `'image'`. Therefore, is this test not already covered by the default case?
Either way it is worth considering adding a test case for `use_mean='none'`

File: digits/model/images/generic/views.py [hunk 7]
```diff
@@ -176,7 +176,7 @@ def generic_image_model_create():
                     val_interval    = form.val_interval.data,
                     pretrained_model= pretrained_model,
                     crop_size       = form.crop_size.data,
-                    use_mean        = bool(form.use_mean.data),
+                    use_mean        = form.use_mean.data,
```
Comment (gheinrich) [inline digits/model/images/generic/views.py]: would you like to add new tests in `digits/model/images/generic/test_views.py` too?

File: digits/model/images/classification/test_views.py [hunk 8]
```diff
@@ -550,6 +564,52 @@ def test_classify_many_json(self):
...
+            image_path = os.path.join(self.imageset_folder, image_path)
+            with open(image_path,'rb') as infile:
+                # StringIO wrapping is needed to simulate POST file upload.
+                image_upload = (StringIO(infile.read()), 'image.png')
+
+            rv = self.app.post(
+                    '/models/images/classification/classify_one?job_id=%s' % self.model_id,
+                    data = {
+                        'image_file': image_upload,
+                        'use_mean': 'pixel',
```
Comment (gheinrich) [inline digits/model/images/classification/test_views.py]: I don't think this form field `use_mean` is used during inference. Instead we are using whichever mean subtraction method was used during training.
If you want to test training and inference with various mean subtraction methods perhaps you're better off creating new subclasses of `BaseTestCreated` like [there](https://github.com/NVIDIA/DIGITS/blob/master/digits/model/images/classification/test_views.py#L642)?

**Comment by gheinrich** [comment]
My auto-encoder network in Torch is quite sensitive to initialization so I thought this was a good candidate to try the various mean subtraction methods on. Along the way I found another issue in the mean pixel subtraction code. Besides I wasn't entirely happy about the Lua wrapper command-line flags (some were mutually exclusive) so I pushed another commit https://github.com/gheinrich/DIGITS/commit/559f6a5ccab8e3df4de290cf9d37eff0e0fc4bc3.
So using mean image subtraction:
mean-image
Using mean pixel subtraction (note the loss is decreasing less rapidly):
mean-pixel
And using no mean subtraction: the network diverges:
mean-none

File: digits/model/images/classification/test_views.py (hunk 6 above)
Comment (lukeyeager) [inline digits/model/images/classification/test_views.py]: > Either way it is worth considering adding a test case for use_mean='none'

:+1:

Comment (jmancewicz) [inline digits/model/images/classification/test_views.py]: Any objection to leaving the image test in there?

File: digits/model/images/generic/views.py (hunk 7 above)
Comment (jmancewicz) [inline digits/model/images/generic/views.py]: Done.

File: digits/model/images/classification/test_views.py (hunk 6 above)
Comment (gheinrich) [inline digits/model/images/classification/test_views.py]: None from me, I think it's good to explicitly test all three cases

**Comment by gheinrich** [comment]
@jmancewicz regarding the two failures on Travis:

`digits.model.images.classification.test_views.TestTorchCreation.test_classify_one_mean_none`:  hopefully you shouldn't see this again with PR #378 

`digits.model.images.generic.test_views.TestTorchCreation.test_diverging_network`: I think this is an integration issue with a38b9e6 as the version on master has this [code](https://github.com/NVIDIA/DIGITS/blob/master/digits/model/tasks/torch_train.py#L298):
```
assert not('inf' in l or 'nan' in l), 'Network reported %s for training loss. Try decreasing your learning rate.' % l
```

**Comment by jmancewicz** [comment]
@gheinrich I dealt with:
`digits.model.images.classification.test_views.TestTorchCreation.test_classify_one_mean_none` by allowing misclassification when nothing is subtracted.

For:
`digits.model.images.generic.test_views.TestTorchCreation.test_diverging_network`, I see that there are differences from  a38b9e6. I grabbed the changes from gheinrich@559f6a5. and have those versions of
```
digits/model/tasks/torch_train.py
tools/torch/data.lua
tools/torch/main.lua
tools/torch/test.lua
```

**Comment by jmancewicz** [comment]
@gheinrich. comparing `master` and `mean` branch execution of this test (`digits.model.images.classification.test_views.TestTorchCreation.test_classify_one_mean_none`) I tarred up the /tmp job dir for each and saw that everything being passed to `th` was the same.  All the files in the dir and the all but one arg being passed.  The only difference was change from `--subtractMean=yes` to `--subtractMean=image`.

Is there something in gheinrich@559f6a5 that could be causing this?

Does this change what is being subtracted from pixel to image, and could that be the problem?

**Comment by gheinrich** [comment]
@jmancewicz.
> Is there something in gheinrich@559f6a5 that could be causing this?

Indeed on this [comment](https://github.com/NVIDIA/DIGITS/pull/321#issuecomment-149188315) I mentioned a bug I found in the Lua wrapper. To fix the bug I reworked the parameters to the Lua script. After the rework there is a single parameter `--subtractMean` with possible values `image`, `pixel` or `none`. 

> I dealt with: `digits.model.images.classification.test_views.TestTorchCreation.test_classify_one_mean_none` by allowing misclassification when nothing is subtracted

Does Caffe not exhibit the same issue on `subtractMean='none'`?

**Comment by jmancewicz** [comment]
@gheinrich, the `digits.model.images.classification.test_views.TestTorchCreation.test_classify_one_mean_none` would fail for me about half the time.  So, it might fail at some point with caffe, but if it does, I haven't seen.  A few days ago, I had many errors, and one may have been a caffe failure on `subtractMean=none`.  @lukeyeager doesn't feel that it needs to classify on that test without the subtraction.  The test is validating that DIGITS doesn't throw any errors.  So, I allowed that test to misclassify and still pass.

It looks like I mentioned the wrong test in the previous comment.  The only one still failing is `digits.model.images.generic.test_views.TestTorchCreation.test_diverging_network`  That's the one that I was a/b-ing with the two tarred directories.  Everything in that comparison is the same except the change from `--subtractMean=yes` to `--subtractMean=image`.  The only other culprit seems to be in  gheinrich@559f6a5. 559f6a5 does have the change to `--subtractMean=image/pixel/none`.

File: digits/model/tasks/torch_train.py [hunk 9]
```diff
@@ -295,7 +297,7 @@ def process_output(self, line):
         if match:
             index = float(match.group(1))
             l = match.group(2)
-            assert not('inf' in l or 'nan' in l), 'Network reported %s for training loss. Try decreasing your learning rate.'  % l
+            assert not('inf' in l or 'nan' in l), 'Network reported %s for training loss. Try changing your learning rate.'  % l
```
Comment (gheinrich) [inline digits/model/tasks/torch_train.py]: You should remove this diff.
`digits.model.images.generic.test_views.TestTorchCreation.test_diverging_network` is failing because the test expects `Try decreasing your learning rate` but here DIGITS throws `Try changing your learning rate`.

File: digits/model/tasks/torch_train.py [hunk 10]
```diff
@@ -122,6 +121,9 @@ def task_arguments(self, resources, env):
         else:
             torch_bin = os.path.join(config_value('torch_root'), 'bin', 'th')
 
+        if self.batch_size is None:
+            self.batch_size = constants.DEFAULT_BATCH_SIZE
+
```
Comment (gheinrich) [inline digits/model/tasks/torch_train.py]: this has been removed on master so I think you need to remove this diff

File: digits/model/tasks/torch_train.py [hunk 11]
```diff
@@ -133,14 +135,12 @@ def task_arguments(self, resources, env):
                 '--save=%s' % self.job_dir,
                 '--snapshotPrefix=%s' % self.snapshot_prefix,
                 '--snapshotInterval=%s' % self.snapshot_interval,
+                '--batchSize=%d' % self.batch_size,
```
Comment (gheinrich) [inline digits/model/tasks/torch_train.py]: this has also been removed on master, you need to keep the code from https://github.com/NVIDIA/DIGITS/blob/master/digits/model/tasks/torch_train.py#L141-L142

File: digits/standard-networks/torch/ImageNet-Training/alexnet.lua [hunk 12]
```diff
@@ -74,9 +74,7 @@ return function(params)
     assert(params.ngpus<=1, 'Model supports only one GPU')
     return {
         model = createModel(1),
-        croplen = 224,
-        trainBatchSize = 100,
-        validationBatchSize = 100,
+        croplen = 224
```
Comment (gheinrich) [inline digits/standard-networks/torch/ImageNet-Training/alexnet.lua]: you need to keep the version on master here

File: digits/standard-networks/torch/ImageNet-Training/googlenet.lua [hunk 13]
```diff
@@ -121,9 +121,7 @@ return function(params)
     assert(params.ngpus<=1, 'Model supports only one GPU')
     return {
         model = createModel(1),
-        croplen = 224,
-        trainBatchSize = 24,
-        validationBatchSize = 24,
+        croplen = 224
```
Comment (gheinrich) [inline digits/standard-networks/torch/ImageNet-Training/googlenet.lua]: you need to keep the version from master here too

File: digits/standard-networks/torch/lenet.lua [hunk 14]
```diff
@@ -36,9 +36,7 @@ return function(params)
     assert(params.ngpus<=1, 'Model supports only CPU or single-GPU')
     return {
         model = lenet,
-        loss = nn.ClassNLLCriterion(),
-        trainBatchSize = 64,
-        validationBatchSize = 100,
+        loss = nn.ClassNLLCriterion()
```
Comment (gheinrich) [inline digits/standard-networks/torch/lenet.lua]: you need to keep the version from master here too

File: docs/InstallTorch.md [hunk 15]
```diff
@@ -82,14 +82,12 @@ ngpus           | number   | Tells how many GPUs are available (0 means CPU)
...
-labelHook             | function     | No        | A function(input,dblabel) that returns the intended label(target) for the current batch given the provided input and label in database. By default the database label is used.
-trainBatchSize        | number       | No        | If specified, sets train batch size. May be overridden by user in DIGITS UI.
-validationBatchSize   | number       | No        | If specified, sets validation batch size. May be overridden by user in DIGITS UI.
+Parameter name  | Type         | Mandatory | Description
+--------------- | ------------ | --------- | -------------
+model           | nn.module    | Yes       | A nn.module container that defines the model to use
+loss            | nn.criterion | No        | A nn.criterion to use during training. Defaults to nn.ClassNLLCriterion.
+croplen         | number       | No        | If specified, inputs images will be cropped randomly to a square of the specified size
+labelHook       | function     | No        | A function(input,dblabel) that returns the intended label(target) for the current batch given the provided input and label in database. By default the database label is used.
```
Comment (gheinrich) [inline docs/InstallTorch.md]: you need to keep the version from master here too

File: digits/model/tasks/torch_train.py (hunk 11 above)
Comment (jmancewicz) [inline digits/model/tasks/torch_train.py]: @gheinrich, I was also missing the lines
```python
            # input images have been resized to network input dimensions by caller
            args.append('--crop=no')
```
Put these back or no?

Comment (gheinrich) [inline digits/model/tasks/torch_train.py]: Yes, well spotted!

**Comment by jmancewicz** [comment]
@gheinrich, I have made the changes from your comments.  Now, it's failing on `digits.model.images.classification.test_views.TestTorchCreation.test_classify_one_mean_pixel` 8 out of 10 times with `AssertionError: image misclassified`.

 The command that is being issued is 
```/Users/jmancewicz/dev/torch/install/bin/th /Users/jmancewicz/dev/digits/DIGITS/tools/torch/test.lua --image=/tmp/tmpGAxJz4.jpeg --network=model --networkDirectory=/tmp/tmpS9N1Uk/20151022-145441-fce9 --load=/tmp/tmpS9N1Uk/20151022-145441-fce9 --snapshotPrefix=snapshot --labels=/tmp/tmpS9N1Uk/20151022-145438-7fb8/labels.txt --mean=/tmp/tmpS9N1Uk/20151022-145438-7fb8/mean.jpg --allPredictions=no --subtractMean=pixel
```

So `--subtractMean=pixel` is being passed.  If it's expected that that should fail, I can ignore the classification error.

File: tools/torch/data.lua [hunk 16]
```diff
@@ -216,9 +212,9 @@ function DBSource:new (backend, db_path, labels_db_path, mirror, crop, croplen,
         self.dbs = {}
         self.total = 0
         for line in file:lines() do
-            local fileName = paths.concat(db_path, paths.basename(line))
+            db_path = paths.concat(db_path, paths.basename(line))
```
Comment (gheinrich) [inline tools/torch/data.lua]: I think you have overwritten the master version here and one line 221

File: tools/torch/data.lua [hunk 17]
```diff
@@ -227,7 +223,7 @@ function DBSource:new (backend, db_path, labels_db_path, mirror, crop, croplen,
             myFile:close()
             -- store DB info
             self.dbs[#self.dbs + 1] = {
-                path = fileName,
+                path = db_path,
```
Comment (gheinrich) [inline tools/torch/data.lua]: here too

File: tools/torch/main.lua [hunk 18]
```diff
@@ -197,19 +184,6 @@ if not opt.crop and network.croplen then
     opt.croplen = network.croplen
 end
 
--- unless specified on command line, inherit train and validation batch size from network
```
Comment (gheinrich) [inline tools/torch/main.lua]: I think you have accidentally removed the following lines

File: tools/torch/main.lua [hunk 19]
```diff
@@ -418,7 +401,7 @@ logmessage.display(0,'initializing the parameters for learning rate policy: ' ..
 local lrpolicy = {}
 if opt.policy ~= 'torch_sgd' then
 
-    local max_iterations = (math.ceil(trainSize/trainBatchSize))*opt.epoch
+    local max_iterations = (math.ceil(trainSize/opt.batchSize))*opt.epoch
```
Comment (gheinrich) [inline tools/torch/main.lua]: this overwrites a change from master too

File: tools/torch/main.lua [hunk 20]
```diff
@@ -542,7 +525,7 @@ end
 
 -- epoch value will be calculated for every batch size. To maintain unique epoch value between batches, it needs to be rounded to the required number of significant digits.
 local epoch_round = 0 -- holds the required number of significant digits for round function.
-local tmp_batchsize = trainBatchSize
+local tmp_batchsize = opt.batchSize
```
Comment (gheinrich) [inline tools/torch/main.lua]: here too (and lines 601, 606, 648, 651, 669, 705)

**Comment by gheinrich** [comment]
Joe, if you wish I can try an integrate my patch again over 84653db?

**Comment by jmancewicz** [comment]
@gheinrich, if this doesn't look correct, then maybe it would be best for you to integrate your patch.  At some point my branch got out of sync, and it's become difficult to know which changes to use from master.

**Comment by gheinrich** [comment]
@jmancewicz I have pushed this [change](https://github.com/NVIDIA/DIGITS/commit/fec4b037e02bb14b1ead95afbe230803ade0b653).
The [Travis test](https://travis-ci.org/NVIDIA/DIGITS/builds/87026952) would have passed if it wasn't for this strange (unrelated?) error (another code -11!) on:
```
ERROR: test suite for <class 'digits.model.images.classification.test_views.TestCaffeCreatedCropInForm'>
...
digits.webapp: ERROR: Create DB (val) task failed with error code -11
```

The C standard error `code -11` means `EAGAIN: try again`...

File: digits/model/tasks/torch_train.py [hunk 21]
```diff
@@ -513,27 +514,22 @@ def classify_one(self, image, snapshot_epoch=None, layers=None):
...
 
         if snapshot_epoch:
             args.append('--epoch=%d' % int(snapshot_epoch))
-        if TORCH_USE_MEAN_PIXEL:
-            args.append('--useMeanPixel=yes')
         if self.trained_on_cpu:
             args.append('--type=float')
 
-        # input image has been resized to network input dimensions by caller
-        args.append('--crop=no')
```
Comment (gheinrich) [inline digits/model/tasks/torch_train.py]: that `--crop=no` has disappeared but it isn't an issue because that flag defaults to `no`

**Comment by gheinrich** [comment]
The changes look OK to me.
